src/lib/SX1280Driver/SX1280_hal.cpp

```diff
@@ -0,0 +1,366 @@
+
+void (*SX1280Hal::TXdoneCallback)() = &nullCallback;
+void (*SX1280Hal::RXdoneCallback)() = &nullCallback;
+
+SX1280Hal::SX1280Hal()
+{
+    instance = this;
+}
+
+void SX1280Hal::end()
+{
+    SPI.end();
+    detachInterrupt(GPIO_PIN_DIO0);
+}
+
+void SX1280Hal::init()
+{
+    Serial.println("Hal Init");
+    pinMode(GPIO_PIN_BUSY, INPUT);
+    pinMode(GPIO_PIN_DIO1, INPUT);
+
+    pinMode(GPIO_PIN_RST, OUTPUT);
+    pinMode(GPIO_PIN_NSS, OUTPUT);
+
+#if defined(GPIO_PIN_RX_ENABLE) || defined(GPIO_PIN_TX_ENABLE)
+    Serial.print("This Target uses seperate TX/RX enable pins: ");
+#endif
+
+#if defined(GPIO_PIN_TX_ENABLE)
+    Serial.print("TX: ");
+    Serial.print(GPIO_PIN_TX_ENABLE);
+#endif
+
+#if defined(GPIO_PIN_RX_ENABLE)
+    Serial.print(" RX: ");
+    Serial.println(GPIO_PIN_RX_ENABLE);
+#endif
+
+#if defined(GPIO_PIN_RX_ENABLE)
+    pinMode(GPIO_PIN_RX_ENABLE, OUTPUT);
+    digitalWrite(GPIO_PIN_RX_ENABLE, LOW);
+#endif
+
+#if defined(GPIO_PIN_TX_ENABLE)
+    pinMode(GPIO_PIN_TX_ENABLE, OUTPUT);
+    digitalWrite(GPIO_PIN_TX_ENABLE, LOW);
+#endif
+
+#ifdef PLATFORM_ESP32
+    SPI.begin(GPIO_PIN_SCK, GPIO_PIN_MISO, GPIO_PIN_MOSI, -1); // sck, miso, mosi, ss (ss can be any GPIO)
+    SPI.setFrequency(18000000);
+#endif
+
+#ifdef PLATFORM_ESP8266
+    Serial.println("PLATFORM_ESP8266");
+    SPI.begin();
+    //SPI.pins(this->SX1280_SCK, this->SX1280_MISO, this->SX1280_MOSI, -1);
+    SPI.setBitOrder(MSBFIRST);
+    SPI.setDataMode(SPI_MODE0);
+    SPI.setFrequency(18000000);
+#endif
+
+#ifdef PLATFORM_STM32
+    SPI.setMOSI(GPIO_PIN_MOSI);
+    SPI.setMISO(GPIO_PIN_MISO);
+    SPI.setSCLK(GPIO_PIN_SCK);
+    SPI.setBitOrder(MSBFIRST);
+    SPI.setDataMode(SPI_MODE0);
+    SPI.begin();
+    SPI.setClockDivider(SPI_CLOCK_DIV4); // 72 / 8 = 9 MHz
+#endif
+
+    //attachInterrupt(digitalPinToInterrupt(GPIO_PIN_BUSY), this->busyISR, CHANGE); //not used atm
+    attachInterrupt(digitalPinToInterrupt(GPIO_PIN_DIO1), this->dioISR, RISING);
+}
+
+void ICACHE_RAM_ATTR SX1280Hal::reset(void)
+{
+    Serial.println("SX1280 Reset");
+    delay(50);
+    digitalWrite(GPIO_PIN_RST, LOW);
+    delay(50);
+    digitalWrite(GPIO_PIN_RST, HIGH);
+
+    while (digitalRead(GPIO_PIN_BUSY) == HIGH) // wait for busy
+    {
+#ifdef PLATFORM_STM32
+        __NOP();
+#elif PLATFORM_ESP32
+        _NOP();
+#elif PLATFORM_ESP8266
+        _NOP();
+#endif
+    }
+
+    //this->BusyState = SX1280_NOT_BUSY;
+    Serial.println("SX1280 Ready!");
+}
+
+void ICACHE_RAM_ATTR SX1280Hal::WriteCommand(SX1280_RadioCommands_t command, uint8_t val)
+{
+    WaitOnBusy();
+    digitalWrite(GPIO_PIN_NSS, LOW);
+
+    SPI.transfer((uint8_t)command);
+    SPI.transfer(val);
+
+    digitalWrite(GPIO_PIN_NSS, HIGH);
+}
+
+void ICACHE_RAM_ATTR SX1280Hal::WriteCommand(SX1280_RadioCommands_t command, uint8_t *buffer, uint8_t size)
+{
+    WORD_ALIGNED_ATTR uint8_t OutBuffer[size + 1];
+
+    OutBuffer[0] = (uint8_t)command;
+    memcpy(OutBuffer + 1, buffer, size);
```

Comment:
Should this be size + 1?

Reply:
no because the command is the first byte and we're copying just the data part